Pickler crash when binding typed pattern to name

## Compiler version

3.3.1-RC1-bin-20230204-a356581-NIGHTLY nad before

## Minimized code

```scala
import scala.quoted.*

case class Foo(x: Int)

inline def foo = ${ fooImpl }

def fooImpl(using Quotes) = 
  val tmp = '{
    1 match
      case x @ (y: Int) => 0
  }

  '{}
```

```scala
val x = foo
```

## Output (click arrow to expand)
<details>

```scala
error] ./MacroTest.scala:1:9: Exception occurred while executing macro expansion.
[error] java.lang.AssertionError: assertion failed: no symbol at Addr(13)
[error]         at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$OwnerTree.search$1(TreeUnpickler.scala:1623)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$OwnerTree.search$1(TreeUnpickler.scala:1629)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$OwnerTree.findOwner(TreeUnpickler.scala:1638)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.symbolAt(TreeUnpickler.scala:268)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readSymRef(TreeUnpickler.scala:261)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readSimpleType$1(TreeUnpickler.scala:423)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readType(TreeUnpickler.scala:464)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthTerm$1(TreeUnpickler.scala:1387)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTerm(TreeUnpickler.scala:1449)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readCase(TreeUnpickler.scala:1502)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readCases$$anonfun$2(TreeUnpickler.scala:1493)
[error]         at dotty.tools.tasty.TastyReader.collectWhile(TastyReader.scala:137)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readCases(TreeUnpickler.scala:1494)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthTerm$1(TreeUnpickler.scala:1337)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTerm(TreeUnpickler.scala:1449)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthTerm$1(TreeUnpickler.scala:1314)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTerm(TreeUnpickler.scala:1449)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTerm(TreeUnpickler.scala:1166)
[error]         at dotty.tools.dotc.core.tasty.TreeUnpickler.unpickle(TreeUnpickler.scala:113)
[error]         at dotty.tools.dotc.core.tasty.DottyUnpickler.computeRootTrees(DottyUnpickler.scala:62)
[error]         at dotty.tools.dotc.ast.tpd$TreeProvider.rootTrees(tpd.scala:1287)
[error]         at dotty.tools.dotc.ast.tpd$TreeProvider.rootTrees$(tpd.scala:1276)
[error]         at dotty.tools.dotc.core.tasty.DottyUnpickler.rootTrees(DottyUnpickler.scala:44)
[error]         at dotty.tools.dotc.ast.tpd$TreeProvider.tree(tpd.scala:1291)
[error]         at dotty.tools.dotc.ast.tpd$TreeProvider.tree$(tpd.scala:1276)
[error]         at dotty.tools.dotc.core.tasty.DottyUnpickler.tree(DottyUnpickler.scala:44)
[error]         at dotty.tools.dotc.quoted.PickledQuotes$.unpickle(PickledQuotes.scala:274)
[error]         at dotty.tools.dotc.quoted.PickledQuotes$.unpickleTerm(PickledQuotes.scala:84)
[error]         at scala.quoted.runtime.impl.QuotesImpl.unpickleExprV2(QuotesImpl.scala:3084)
[error]         at Macro$package$.fooImpl(Macro.scala:8)
[error] 
[error] val x = foo
[error]         ^^^
```
</details>

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
@@ -74,6 +74,9 @@ class TreeUnpickler(reader: TastyReader,
    */
   private val typeAtAddr = new mutable.HashMap[Addr, Type]
 
+  /** If this is a pickled quote, the owner of the quote, otherwise NoSymbol. */
+  private var rootOwner: Symbol = NoSymbol
+
   /** The root symbol denotation which are defined by the Tasty file associated with this
    *  TreeUnpickler. Set by `enterTopLevel`.
    */
@@ -106,6 +109,7 @@ class TreeUnpickler(reader: TastyReader,
 
   /** The unpickled trees */
   def unpickle(mode: UnpickleMode)(using Context): List[Tree] = {
+    if mode != UnpickleMode.TopLevel then rootOwner = ctx.owner
     assert(roots != null, "unpickle without previous enterTopLevel")
     val rdr = new TreeReader(reader)
     mode match {
@@ -1635,7 +1639,7 @@ class TreeUnpickler(reader: TastyReader,
             pickling.println(i"no owner for $addr among $cs%, %")
             throw ex
         }
-      try search(children, NoSymbol)
+      try search(children, rootOwner)
       catch {
         case ex: TreeWithoutOwner =>
           pickling.println(s"ownerTree = $ownerTree")

--- a/tests/pos-macros/i16843a/Macro_1.scala
+++ b/tests/pos-macros/i16843a/Macro_1.scala
@@ -1,0 +1,13 @@
+import scala.quoted.*
+
+case class Foo(x: Int)
+
+inline def foo = ${ fooImpl }
+
+def fooImpl(using Quotes) =
+  val tmp = '{
+    1 match
+      case x @ (y: Int) => 0
+  }
+
+  '{}

--- a/tests/pos-macros/i16843a/Test_2.scala
+++ b/tests/pos-macros/i16843a/Test_2.scala
@@ -1,0 +1,1 @@
+val x = foo

--- a/tests/pos-macros/i16843b/Macro_1.scala
+++ b/tests/pos-macros/i16843b/Macro_1.scala
@@ -1,0 +1,18 @@
+import scala.quoted.*
+
+inline def foo: Int = ${ fooImpl }
+
+def fooImpl(using Quotes): Expr[Int] =
+  '{
+    val b = ${
+      val a = '{
+        (1: Int) match
+          case x @ (y: Int) => 0
+      }
+      a
+    }
+
+    (1: Int) match
+      case x @ (y: Int) => 0
+  }
+

--- a/tests/pos-macros/i16843b/Test_2.scala
+++ b/tests/pos-macros/i16843b/Test_2.scala
@@ -1,0 +1,1 @@
+def test = foo

